File: mami.r
mami <- function(train, test, train_labels, k1=5, k2=10){

D  = rbind(train, test)
D1 = as.matrix(dist(scale(D)))
#D1 = as.matrix(dist(D))

# Calculate first layer neigborhood for each test point
K1 = matrix(NaN,nrow(test),k1)
K1_l = matrix(NaN,nrow(test),k1)
count = 1 
for (xx in (nrow(train)+1):(nrow(train)+nrow(test))){

    vec = D1[xx,1:nrow(train)]
    vec_s = sort(vec)
    K1_l[count,] = train_labels[as.numeric(names(head(vec_s,k1)))]
    K1[count,] = as.numeric(names(head(vec_s,k1)))
    count = count + 1
    #print(vec_s)

}
#print(K1_l)
#print(K1)

# Calculate Second layer Neighborhood for each point 
K2 = matrix(NaN, k1, k2)
K2_list = list()

for(xx in 1:nrow(K1)){
    for(yy in 1:k1){
        vec = D1[K1[xx,yy],1:nrow(train)]
        vec_s = sort(vec)
        K2[yy,] = train_labels[as.numeric(names(head(vec_s,k2)))]
    }
  K2_list[[xx]] = K2
  K2 = matrix(NaN, k1, k2)
}
#print(K2_list)

C_list = list()
C = rep(NaN, k1)
# Cac Coverage
for(xx in 1:length(K2_list)){
    for(yy in 1:k1){
        C[yy] = sum(K2_list[[xx]][yy,]==K1_l[xx,yy])         
    }
    C_list[[xx]] = C
    C = rep(NaN, k1)
}
#print(C_list)

COV_ALL = list()

# Now infer the prediction
for(xx in 1:nrow(K1_l)){
    l1 = table(K1_l[xx,])
    maj = as.numeric(names(l1)[1])
    COV = rep(NaN, length(l1))
    names(COV) = names(l1)
    for(yy in 1:length(l1)){
        ids = which(K1_l[xx,] == as.numeric(names(l1[yy])))
        #ids = which(K1_l[xx,] == maj)
        #COV[yy] = sum(C_list[[xx]][ids])/(length(ids)*k2)
        COV[yy] = sum(C_list[[xx]][ids])/(k1*k2)
    }

    #print(COV)
  COV_ALL[[xx]] = COV 
}

#print(COV_ALL)

PRED = as.numeric(sapply(COV_ALL, function(x){names(which.max(x))}))

COV = matrix(0, length(PRED), length(unique(train_labels)))
for(xx in 1:dim(COV)[1]){

    ids = as.numeric(names(COV_ALL[[xx]])) #+ 1
    #print(ids)
    #print(COV_ALL[[xx]])
    COV[xx, ids] = COV_ALL[[xx]]

}


#COV  = as.numeric(sapply(COV_ALL, function(x){max(x)}))

# NORMALIZE
#COV <- t(apply(COV, 1, function(row) row / sum(row)))

return(list(prediction=PRED, coverage=COV))

}

File: mami.py
import numpy as np
from scipy.spatial.distance import cdist
from sklearn.preprocessing import StandardScaler

def mami_py(train, test, train_labels, k1=5, k2=10, 
         distance='euclidean', scaling=True, dmatrix=None):
    
    n_train = train.shape[0]
    n_test = test.shape[0]
    all_data = np.vstack([train, test])

    # Compute distance matrix if not provided
    if dmatrix is None:
        if scaling:
            all_data = StandardScaler().fit_transform(all_data)
        D1 = cdist(all_data, all_data, metric=distance)
    else:
        D1 = dmatrix

    # First-layer neighbors
    K1 = np.full((n_test, k1), np.nan)
    K1_l = np.full((n_test, k1), np.nan)

    for i in range(n_test):
        test_idx = n_train + i
        dists = D1[test_idx, :n_train]
        nearest = np.argsort(dists)[:k1]
        K1[i, :] = nearest
        K1_l[i, :] = train_labels[nearest]

    # Second-layer neighbors
    K2_list = []

    for i in range(n_test):
        K2 = np.full((k1, k2), np.nan)
        for j in range(k1):
            idx = int(K1[i, j])
            dists = D1[idx, :n_train]
            nearest = np.argsort(dists)[:k2]
            K2[j, :] = train_labels[nearest]
        K2_list.append(K2)

    # Coverage calculation
    C_list = []
    for i in range(n_test):
        C = np.zeros(k1)
        for j in range(k1):
            C[j] = np.sum(K2_list[i][j, :] == K1_l[i, j])
        C_list.append(C)

    # Predict and compute coverage matrix
    PRED = []
    COV_ALL = []

    for i in range(n_test):
        labels, counts = np.unique(K1_l[i, :], return_counts=True)
        COV = {}
        for lbl in labels:
            mask = K1_l[i, :] == lbl
            COV[str(int(lbl))] = np.sum(C_list[i][mask]) / (k1 * k2)
        COV_ALL.append(COV)
        pred_label = int(max(COV, key=COV.get))
        PRED.append(pred_label)

    # Build final coverage matrix
    unique_labels = np.unique(train_labels)
    label_to_idx = {str(int(lbl)): i for i, lbl in enumerate(unique_labels)}
    COV = np.zeros((n_test, len(unique_labels)))

    for i, cov_dict in enumerate(COV_ALL):
        for lbl_str, val in cov_dict.items():
            j = label_to_idx[lbl_str]
            COV[i, j] = val

    return {
        'prediction': np.array(PRED),
        'coverage': COV
    }